### PR TITLE
Enable logical planner for single-shard tables

### PR DESCRIPTION
**Pull request by onurctirtir**

### CODE REVIEW

File: src/test/regress/expected/query_single_shard_table.out
Comment (onurctirtir) [inline src/test/regress/expected/query_single_shard_table.out:914]: We could be smarter than this but I feel like this is fine for 12.0

File: src/backend/distributed/planner/query_colocation_checker.c
Comment (onurctirtir) [inline src/backend/distributed/planner/query_colocation_checker.c:174]: should we be smarter when choosing anchor rel?

File: src/backend/distributed/planner/relation_restriction_equivalence.c
Comment (onderkalaci) [inline src/backend/distributed/planner/relation_restriction_equivalence.c:2007]: I cannot follow this logic. Why do we decide to allow two append tables to be colocated now?

Comment (onurctirtir) [reply src/backend/distributed/planner/relation_restriction_equivalence.c:2007]: As before, we say that two append tables are not-colocated. However before, we were saying that the query references non-colocated append tables even if it only references a single append table and this PR fixes that. This wasn't a problem before but now we're calling this function in more places, hence needed to fix that.

Comment (onderkalaci) [reply src/backend/distributed/planner/relation_restriction_equivalence.c:2007]: > As before, we say that two append tables are not-colocated

 the code doesn't seem to do that anymore? Before this patch, it was impossible for any two append tables to be decided as colocated, now it is possible?

> non-colocated append

append tables are always non-colocated, so I cannot follow this

> if it only references a single append table

Well, then we should make that more generic, at the top of the function. If there is a single table, we always return true. Having a special case for append is confusing 

> SELECT key, value FROM append_table a WHERE key IN (SELECT key FROM append_table WHERE value > 100) ORDER BY 1,2;

This query now starts to work. Is this code change allowing it? Can you share the explain plan of this query?

File: src/backend/distributed/planner/query_colocation_checker.c
Comment (onderkalaci) [inline src/backend/distributed/planner/query_colocation_checker.c:189]: why do we continue? The currentRTE is a distributed table with dist key, and shouldn't we prefer that becoming the `anchor` rte? Otherwise:
```SQL
explain select * from (SELECT * FROM nullkey_c1_t1) foo JOIN (SELECT * FROM distributed_table) as bar USING (a) JOIN (SELECT * FROM distributed_table) as foo2 USING (a) LIMIT 1;
┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                            QUERY PLAN                                                             │
├───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=0 width=0)                                                                    │
│   ->  Distributed Subplan 44_1                                                                                                    │
│         ->  Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=100000 width=8)                                                   │
│               Task Count: 32                                                                                                      │
│               Tasks Shown: One of 32                                                                                              │
│               ->  Task                                                                                                            │
│                     Node: host=localhost port=9701 dbname=onderkalaci                                                             │
│                     ->  Seq Scan on distributed_table_102014 distributed_table  (cost=0.00..32.60 rows=2260 width=8)              │
│   ->  Distributed Subplan 44_2                                                                                                    │
│         ->  Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=100000 width=8)                                                   │
│               Task Count: 32                                                                                                      │
│               Tasks Shown: One of 32                                                                                              │
│               ->  Task                                                                                                            │
│                     Node: host=localhost port=9701 dbname=onderkalaci                                                             │
│                     ->  Seq Scan on distributed_table_102014 distributed_table  (cost=0.00..32.60 rows=2260 width=8)              │
│   Task Count: 1                                                                                                                   │
│   Tasks Shown: All                                                                                                                │
│   ->  Task                                                                                                                        │
│         Node: host=localhost port=9702 dbname=onderkalaci                                                                         │
│         ->  Limit  (cost=0.01..5.11 rows=1 width=16)                                                                              │
│               ->  Nested Loop  (cost=0.01..16970.17 rows=3325 width=16)                                                           │
│                     Join Filter: (nullkey_c1_t1.a = intermediate_result_1.a)                                                      │
│                     ->  Nested Loop  (cost=0.00..2007.66 rows=665 width=16)                                                       │
│                           Join Filter: (nullkey_c1_t1.a = intermediate_result.a)                                                  │
│                           ->  Function Scan on read_intermediate_result intermediate_result  (cost=0.00..10.00 rows=1000 width=8) │
│                           ->  Materialize  (cost=0.00..3.00 rows=133 width=8)                                                     │
│                                 ->  Seq Scan on nullkey_c1_t1_102008 nullkey_c1_t1  (cost=0.00..2.33 rows=133 width=8)            │
│                     ->  Function Scan on read_intermediate_result intermediate_result_1  (cost=0.00..10.00 rows=1000 width=8)     │
└───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
(28 rows)
```

Overall, I think we really should simplify the logic here. The AnchorRTE selection is NOT optimal anyway, and we are not gaining much here. So, can you just get back to the original code:

```SQL
		else if (currentRte->rtekind == RTE_RELATION)
		{
			Oid relationId = currentRte->relid;

			if (!IsCitusTableType(relationId, DISTRIBUTED_TABLE)
			{
				/*
				 * We're not interested in non distributed relations.
				 */
				continue;
			}

			anchorRangeTblEntry = currentRte;
			break;
		}
```

At somepoint we might optimize this code, for now, lets just keep it simple

Comment (onderkalaci) [reply src/backend/distributed/planner/query_colocation_checker.c:189]: For testing, play with queries in the following form and make sure anything non-colocated gets recursively planned:
```SQL
explain select * from (SELECT * FROM nullkey_c1_t1) foo JOIN (SELECT * FROM distributed_table) as bar USING (a) JOIN (SELECT * FROM distributed_table) as foo2 USING (a) LIMIT 1;

explain select * from nullkey_c1_t1 foo JOIN  distributed_table as bar USING (a) JOIN (SELECT * FROM distributed_table) as foo2 USING (a) LIMIT 1;

```

where change the orders of foo/bar/foo2 and sometimes use them directly as tables sometimes inside subqueries. 

Or, make subqueries non-colocated by using non-dist keys joining -- see `b as a` in the target list:

```SQL
explain select * from (SELECT * FROM nullkey_c1_t1) foo JOIN distributed_table as bar USING (a) JOIN (SELECT b as a FROM distributed_table) as foo2 USING (a) LIMIT 1;
```

File: src/backend/distributed/planner/query_pushdown_planning.c
Comment (onderkalaci) [inline src/backend/distributed/planner/query_pushdown_planning.c:592]: Can you explain the reason why we ended-up adding `AllDistributedRelationsInRestrictionContextColocated` all over the planner, and we expanded equality checks with colocation checks?

I really doubt we need it in so many places.  Can you, for example, share the example queries that are fixed by this?

ps: we already have colocation checks in place, otherwise we'd return wrong results.

File: src/backend/distributed/planner/recursive_planning.c
Comment (onderkalaci) [inline src/backend/distributed/planner/recursive_planning.c:1310]: Again, I cannot follow this here.

We are already capable of doing non-colocated subquery joins without this on non-colocated tables:

```SQL
explain select * from (SELECT * FROM dist_2) foo JOIN distributed_table as bar USING (a) LIMIT 1;
┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                            QUERY PLAN                                                             │
├───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Limit  (cost=0.00..0.00 rows=1 width=8)                                                                                           │
│   ->  Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=100000 width=8)                                                         │
│         ->  Distributed Subplan 12_1                                                                                              │
│               ->  Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=100000 width=4)                                             │
│                     Task Count: 32                                                                                                │
│                     Tasks Shown: One of 32                                                                                        │
│                     ->  Task                                                                                                      │
│                           Node: host=localhost port=9701 dbname=onderkalaci                                                       │
│                           ->  Seq Scan on dist_2_102047 dist_2  (cost=0.00..35.50 rows=2550 width=4)                              │
│         Task Count: 32                                                                                                            │
│         Tasks Shown: One of 32                                                                                                    │
│         ->  Task                                                                                                                  │
│               Node: host=localhost port=9701 dbname=onderkalaci                                                                   │
│               ->  Limit  (cost=0.00..3.01 rows=1 width=8)                                                                         │
│                     ->  Nested Loop  (cost=0.00..33948.25 rows=11300 width=8)                                                     │
│                           Join Filter: (intermediate_result.a = bar.a)                                                            │
│                           ->  Function Scan on read_intermediate_result intermediate_result  (cost=0.00..10.00 rows=1000 width=4) │
│                           ->  Materialize  (cost=0.00..43.90 rows=2260 width=8)                                                   │
│                                 ->  Seq Scan on distributed_table_102014 bar  (cost=0.00..32.60 rows=2260 width=8)                │
└───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
(19 rows)
```

where
```SQL
 select colocationid from pg_dist_partition WHERE logicalrelid IN ('distributed_table'::regclass, 'dist_2'::regclass);
┌──────────────┐
│ colocationid │
├──────────────┤
│            5 │
│            6 │
└──────────────┘
(2 rows)
```

So, overall, if you can point me the reason why we need these colocation checks, I'd be happy to provide input. But, my current thinking is that we should most likely undo all of these.

File: src/backend/distributed/planner/query_colocation_checker.c
Comment (onderkalaci) [reply src/backend/distributed/planner/query_colocation_checker.c]: Some other example tests cases:
https://github.com/citusdata/citus/blob/main/src/test/regress/sql/having_subquery.sql
https://github.com/citusdata/citus/blob/main/src/test/regress/sql/subquery_and_cte.sql
https://github.com/citusdata/citus/commit/f2538a456f075f15e475e87260c99062992dd69d#diff-db780b730b14664acd9e7ae3c0965f35011274cc25c6114f812467198b0ca921R5

Comment (onurctirtir) [reply src/backend/distributed/planner/query_colocation_checker.c]: * try a few different arbitrary config tests, such as single node?
* update delete using complex subqueries

File: src/backend/distributed/planner/relation_restriction_equivalence.c
Comment (onurctirtir) [reply src/backend/distributed/planner/relation_restriction_equivalence.c]: > I cannot follow this logic. Why do we decide to allow two append tables to be colocated now?

As we've discussed, this change has been reverted.

File: src/backend/distributed/planner/insert_select_planner.c
Comment (onurctirtir) [reply src/backend/distributed/planner/insert_select_planner.c:1408]: Brought the assert back.

> But for both of the above, would be nice to open an issue afterwards saying that INSERT .. SELECT repartitioning is not supported for single shard distributed tables neither for source/target.

I took a note to document this in the parent issue of single-shard tables at the end of the release.

File: src/backend/distributed/planner/recursive_planning.c
Comment (onurctirtir) [reply src/backend/distributed/planner/recursive_planning.c]: As we've discussed, this change has been reverted.

File: src/backend/distributed/planner/query_pushdown_planning.c
Comment (onurctirtir) [reply src/backend/distributed/planner/query_pushdown_planning.c]: As we've discussed, this change has been reverted.

File: src/test/regress/expected/query_single_shard_table.out
Comment (onderkalaci) [inline src/test/regress/expected/query_single_shard_table.out:481]: maybe play with citus.local_table_join_policy  for 1 test

File: src/test/regress/expected/join_pushdown.out
Comment (onderkalaci) [inline src/test/regress/expected/join_pushdown.out:155]: maybe run https://github.com/citusdata/citus/tree/main/src/test/regress/citus_tests/query_generator on null-shard-key tables

File: src/test/regress/citus_tests/config.py
Comment (onurctirtir) [reply src/test/regress/citus_tests/config.py:219]: > -- nested execution is allowed outside of an aggregate
-- note that this behaviour is different if distributed has only 1 shard
-- however, this test always uses 4 shards
SELECT dist_query_single_shard(count(*)::int) FROM distributed;

File: src/test/regress/expected/query_single_shard_table.out
Comment (onurctirtir) [reply src/test/regress/expected/query_single_shard_table.out:1948]: As we've discussed in group chat, resolving this as we're ok with that. I'll document this in the parent issue if time doesn't permit for this release.

**Review by onderkalaci — COMMENTED** [review]
For reference, I tried:
- repartition joins (including reference tables in the tree)
- recursive planning (including non-colocated subqueies, having subqueries, outer joins, local table joins, set operations etc.)
- router queries works fine already
- INSERT .. SELECT combinations
- UPDATE/DELETE combinations

Still, let me approve once you add some tests and check my comments

**Review by onderkalaci — COMMENTED** [review]
For reference, I tried:
- repartition joins (including reference tables in the tree)
- recursive planning (including non-colocated subqueies, having subqueries, outer joins, local table joins etc.)
- router queries works fine already
- INSERT .. SELECT combinations
- UPDATE/DELETE combinations

Still, let me approve once you add some tests and check my comments

File: src/backend/distributed/planner/query_colocation_checker.c
Comment (onurctirtir) [reply src/backend/distributed/planner/query_colocation_checker.c]: Note to myself: only remaining thing to check is the last one:
> https://github.com/citusdata/citus/commit/f2538a456f075f15e475e87260c99062992dd69d#diff-db780b730b14664acd9e7ae3c0965f35011274cc25c6114f812467198b0ca921R5

File: src/test/regress/citus_tests/config.py
Comment (onderkalaci) [reply src/test/regress/citus_tests/config.py:219]: what happens when you enable this test? It does seem to query a table? maybe change the query to `SELECT dist_query_single_shard(count(*)::int) == X ROM distributed;`, where `X` is what changes depending on the shard count?

**Comment by aykut-bozkurt** [comment]
Error found by query generator

```sql
SELECT count(*), avg(avgsub.id) FROM  (  SELECT  table_0.id  FROM  ref1  AS  table_0 INNER JOIN single_dist1  AS  table_1 USING (id )  INNER JOIN ref1  AS  table_2 USING (id )  INNER JOIN single_dist0  AS  table_3 USING (id )  ORDER BY id  LIMIT 7 )  AS avgsub;

ERROR:  relation "public.single_dist1_102022" does not exist
```

**Comment by aykut-bozkurt** [comment]
```sql
SELECT count(*), avg(avgsub.id) FROM  (  SELECT  table_0.id  FROM  single_dist1  AS  table_0 RIGHT JOIN (  SELECT  table_2.id  FROM  (  SELECT  table_3.id  FROM  single_dist0  AS  table_3 ORDER BY id  LIMIT 0 )  AS  table_2 INNER JOIN single_dist0  AS  table_4 USING (id )  WHERE  table_4.id < 8 )  AS  table_1 USING (id )  )  AS avgsub;

ERROR:  cannot perform a lateral outer join when a distributed subquery references complex subqueries, CTEs or local tables
```

The same DDLs above.

File: src/backend/distributed/planner/query_colocation_checker.c
Comment (onurctirtir) [reply src/backend/distributed/planner/query_colocation_checker.c]: done

File: src/test/regress/expected/join_pushdown.out
Comment (onurctirtir) [reply src/test/regress/expected/join_pushdown.out:155]: (@aykut-bozkurt is doing some testing based on query generator, fyi)

**Comment by onderkalaci** [comment]
> ```sql
> SELECT count(*), avg(avgsub.id) FROM  (  SELECT  table_0.id  FROM  ref1  AS  table_0 INNER JOIN single_dist1  AS  table_1 USING (id )  INNER JOIN ref1  AS  table_2 USING (id )  INNER JOIN single_dist0  AS  table_3 USING (id )  ORDER BY id  LIMIT 7 )  AS avgsub;
> ```

Good catch, it seems the ancient  `AnchorRangeTableIdList` function checks shard count as well. And, both reference tables and single shard tables have 1 shard. So, it just picks the anchor relation randomly. We should change that function such that it should dis-favor reference tables.

@onurctirtir can you have a look please?

File: src/test/regress/sql/query_single_shard_table.sql
Comment (onderkalaci) [reply src/test/regress/sql/query_single_shard_table.sql:190]: (for the bug that the tool finds, when you fix it, please add the test with single repartition enabled as well, that might go from a different route)

**Comment by onurctirtir** [comment]
> ```sql
> SELECT count(*), avg(avgsub.id) FROM  (  SELECT  table_0.id  FROM  single_dist1  AS  table_0 RIGHT JOIN (  SELECT  table_2.id  FROM  (  SELECT  table_3.id  FROM  single_dist0  AS  table_3 ORDER BY id  LIMIT 0 )  AS  table_2 INNER JOIN single_dist0  AS  table_4 USING (id )  WHERE  table_4.id < 8 )  AS  table_1 USING (id )  )  AS avgsub;
> 
> ERROR:  cannot perform a lateral outer join when a distributed subquery references complex subqueries, CTEs or local tables
> ```
> 
> The same DDLs above.

This seem to be expected, I got an error (even if it's a different one) if I replace single-shard tables with regular distributed tables:
> ERROR:  recursive complex joins are only supported when all distributed tables are co-located and joined on their distribution columns

**Comment by aykut-bozkurt** [comment]
> > ```sql
> > SELECT count(*), avg(avgsub.id) FROM  (  SELECT  table_0.id  FROM  single_dist1  AS  table_0 RIGHT JOIN (  SELECT  table_2.id  FROM  (  SELECT  table_3.id  FROM  single_dist0  AS  table_3 ORDER BY id  LIMIT 0 )  AS  table_2 INNER JOIN single_dist0  AS  table_4 USING (id )  WHERE  table_4.id < 8 )  AS  table_1 USING (id )  )  AS avgsub;
> > 
> > ERROR:  cannot perform a lateral outer join when a distributed subquery references complex subqueries, CTEs or local tables
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The same DDLs above.
> 
> This seem to be expected, I got an error (even if it's a different one) if I replace single-shard tables with regular distributed tables:
> 
> > ERROR:  recursive complex joins are only supported when all distributed tables are co-located and joined on their distribution columns

Then we can ignore the error similarly at query generator diff checker.

**Review by onderkalaci — APPROVED** [review]
I'm approving a little early for logistic reasons, but expect 2 things before you merge:
- make sure that all the queries you added to the regressions tests gives the exact results on PG. I'd typically re-run the test files by commenting out the create_distributed/reference_table calls. Of course, some modifications queries might work after that, which might change results. I'd comment those out as well.
- Make sure that the query generator tool works smoothly after your fix.

**Comment by onurctirtir** [comment]
> * make sure that all the queries you added to the regressions tests gives the exact results on PG. I'd typically re-run the test files by commenting out the create_distributed/reference_table calls. Of course, some modifications queries might work after that, which might change results. I'd comment those out as well.

Done, wrapped all failing UPDATE/DELETE/INSERT queries with BEGIN/ROLLBACK stmts and ensured that we don't return different results for single shard tables.

> * Make sure that the query generator tool works smoothly after your fix.

As far as I heard from @aykut-bozkurt, we're fine now.

File: src/backend/distributed/planner/multi_physical_planner.c
Comment (onurctirtir) [reply src/backend/distributed/planner/multi_physical_planner.c:2947]: done

File: src/test/regress/citus_tests/config.py
Comment (onurctirtir) [reply src/test/regress/citus_tests/config.py:219]: We've decided to not enable this tests for single shard tables because there are lots of tests in this test file that fail when the table has one shard.

**Pull request merged**
